src/sentry/services/hybrid_cloud/organization_mapping/impl.py

```diff
@@ -0,0 +1,75 @@
+from dataclasses import fields
+from typing import Any, Optional
+
+from django.db import IntegrityError, transaction
+
+from sentry.models.organizationmapping import OrganizationMapping
+from sentry.models.user import User
+from sentry.services.hybrid_cloud.organization_mapping import (
+    APIOrganizationMapping,
+    OrganizationMappingService,
+)
+
+
+class DatabaseBackedOrganizationMappingService(OrganizationMappingService):
+    def create(
+        self,
+        *,
+        user: User,
+        organization_id: int,
+        slug: str,
+        region_name: str,
+        idempotency_key: Optional[str] = "",
+        # There's only a customer_id when updating an org slug
+        customer_id: Optional[str] = None,
+    ) -> APIOrganizationMapping:
+        try:
+            with transaction.atomic():
+                # Creating an identical mapping should succeed, even if a record already exists
+                # with this slug. We allow this IFF the idempotency key is identical.
+                org_mapping = OrganizationMapping.objects.create(
+                    organization_id=organization_id,
+                    slug=slug,
+                    customer_id=customer_id,
+                    idempotency_key=idempotency_key,
+                    region_name=region_name,
+                )
+            return self.serialize_organization_mapping(org_mapping)
+        except IntegrityError:
+            pass
+
+        # If we got here, the slug already exists
+        if idempotency_key != "":
+            try:
+                with transaction.atomic():
+                    existing_mapping = OrganizationMapping.objects.select_for_update().get(
+                        slug=slug, idempotency_key=idempotency_key
+                    )
+                    existing_mapping.update(
+                        organization_id=organization_id,
+                        customer_id=customer_id,
+                        region_name=region_name,
+                    )
+                return self.serialize_organization_mapping(existing_mapping)
+            except OrganizationMapping.DoesNotExist:
+                pass
+
+        raise IntegrityError("An organization with this slug already exists.")
```

Comment:
A very, very important note here: because the service layer will likely become RPC, it's important to remember that *exceptions won't carry*.  In fact, currently, I explicitly catch all exceptions in stubbed services and convert them to general runtime exceptions, because in fact when RPC comes along these calls won't carry their original exception type (only likely a general error message from the service).

Reply:
Great point. Our RPC layer will need to figure this out.

Reply:
Agreed, I'll leave the integrity error for now but we'll need to come back to this when implementing RPC.